crates/chain-listener/src/listener.rs

```diff
@@ -779,65 +791,126 @@ impl ChainListener {
+                    .unwrap_or(false)
+            })
+            .collect()
+    }
+
+    /// Return proofs in CC that is not active yet and can't produce proofs in the current epoch
+    fn get_pending_units(&self) -> Vec<CUID> {
+        self.cc_compute_units
+            .values()
+            .filter(|cu| cu.startEpoch > self.current_epoch)
+            .map(|cu| CUID::new(cu.id.0))
+            .collect()
+    }
+
     /// Send GlobalNonce, Difficulty and Core<>CUID mapping (full commitment info) to CCP
     async fn refresh_commitment(&self) -> eyre::Result<()> {
-        if self.active_compute_units.is_empty() {
+        if self.cc_compute_units.is_empty() || self.current_commitment.is_none() {
             self.stop_commitment().await?;
             return Ok(());
         }
 
         tracing::info!(target: "chain-listener",
             "Refreshing commitment, active compute units: {}",
-            self.active_compute_units
-                .iter()
+            self.cc_compute_units
+                .keys()
                 .map(CUID::to_string)
                 .collect::<Vec<_>>()
                 .join(", ")
         );
         tracing::info!(target: "chain-listener", "Global nonce: {}", self.global_nonce);
         tracing::info!(target: "chain-listener", "Difficulty: {}", self.difficulty);
         if let Some(ref ccp_client) = self.ccp_client {
-            let mut cores = self.acquire_active_units()?;
-
-            // All pending units not involved in deals will help to solve CCs for other units
-            tracing::info!(target: "chain-listener",
-                "Pending compute units: {:?}",
-                self.pending_compute_units
-                    .iter()
-                    .map(|cu| cu.id.to_string())
-                    .collect::<Vec<_>>()
-            );
-            let mut available_cores = self.get_available_cores()?;
-            tracing::info!(
-                target: "chain-listener",
-                "{} cores of pending units will be allocated to other units",
-                available_cores.len()
-            );
-
-            for unit in self.active_compute_units.iter().cycle() {
-                if let Some(core) = available_cores.pop_first() {
-                    cores.insert(core, *unit);
+            let priority_units = self.get_priority_units();
+            let non_priority_units = self.get_non_priority_units();
+            let pending_units = self.get_pending_units();
+
+            let mut cores = HashMap::new();
+            let priority_cores = self.acquire_units(priority_units.clone())?;
+            let non_priority_cores = self.acquire_units(non_priority_units.clone())?;
+            let mut pending_cores = self.acquire_units(pending_units.clone())?;
+
+            if priority_units.is_empty() {
+                tracing::info!(target: "chain-listener", "All CUs found MIN_PROOFS_PER_EPOCH proofs in current epoch {}", self.current_epoch);
```

Comment:
can you print `MIN_PROOFS_PER_EPOCH` value?

```suggestion
                tracing::info!(target: "chain-listener", "All CUs found minimal number of proofs {} in current epoch {}", MIN_PROOF_COUNT, self.current_epoch);
```